File: etl/congresso/etl/temas.py
"""Temas editoriais — espelho Python de `rubrica/temas.json`.

Mesma fonte canônica que `lib/temas.ts` (que faz o filtro na UI). Usado
aqui para SELECIONAR lotes de análise por tema — ver `etl/fila_prioridade.py`.
Nunca duplicar a definição de um tema fora do JSON: os dois lados
divergiriam silenciosamente sobre o que "é" cada tema.
"""
import json
import re
import unicodedata
from pathlib import Path
from typing import Any

TEMAS_PATH = Path(__file__).resolve().parents[3] / "apps" / "web" / "lib" / "congresso" / "rubrica" / "temas.json"
TEMAS: list[dict[str, Any]] = json.loads(TEMAS_PATH.read_text(encoding="utf-8"))["temas"]

_TEMAS_POR_SLUG = {t["slug"]: t for t in TEMAS}


def tema_por_slug(slug: str) -> dict[str, Any] | None:
    return _TEMAS_POR_SLUG.get(slug)


def _normalizar(s: str) -> str:
    return unicodedata.normalize("NFC", s or "")


def casa_com_tema(tema: dict[str, Any], prop: dict, direitos_da_analise: list[str] | None = None) -> bool:
    """Espelha `casaComTema` de lib/temas.ts, campo a campo."""
    direitos_da_analise = direitos_da_analise or []
    oficiais = set(prop.get("temas_oficiais") or [])
    if any(t in oficiais for t in tema.get("temasOficiais", [])):
        return True
    if any(d in direitos_da_analise for d in tema.get("direitos", [])):
        return True
    padrao = tema.get("padrao")
    if padrao:
        alvo = f"{prop.get('ementa') or ''} {prop.get('keywords') or ''}"
        if re.search(padrao, _normalizar(alvo), re.IGNORECASE):
            return True
    return False

File: apps/web/lib/congresso/temas.ts
import temasJson from "./rubrica/temas.json";

/**
 * Temas editoriais do portal — espelho fino de `rubrica/temas.json`.
 *
 * A definição não mora aqui: ver a `_nota`/`_nota_fontes` no JSON para o
 * porquê de três fontes por tema e o histórico de calibração dos regex.
 * `etl/temas.py` lê o MESMO arquivo para selecionar lotes de análise — os
 * dois lados nunca podem divergir na definição de um tema.
 */

export interface Tema {
  slug: string;
  nome: string;
  descricao: string;
  temasOficiais?: string[];
  direitos?: string[];
  padrao?: string;
}

export const TEMAS: Tema[] = temasJson.temas as Tema[];

export function temaPorSlug(slug: string): Tema | undefined {
  return TEMAS.find((t) => t.slug === slug);
}

export function regexDoTema(tema: Tema): RegExp | null {
  return tema.padrao ? new RegExp(tema.padrao, "i") : null;
}

/** Uma proposição pertence ao tema? Avaliado em memória, ver `lib/destaques.ts`. */
export function casaComTema(
  tema: Tema,
  prop: { ementa?: string | null; keywords?: string | null; temas_oficiais?: string[] | null },
  direitosDaAnalise: string[] = []
): boolean {
  if (tema.temasOficiais?.some((t) => (prop.temas_oficiais ?? []).includes(t))) return true;
  if (tema.direitos?.some((d) => direitosDaAnalise.includes(d))) return true;
  const rx = regexDoTema(tema);
  if (rx && rx.test(`${prop.ementa ?? ""} ${prop.keywords ?? ""}`)) return true;
  return false;
}
